x-pack/plugins/infra/public/locators/helpers.ts

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { interpret } from 'xstate';
+import { waitFor } from 'xstate/lib/waitFor';
+import { flowRight } from 'lodash';
+import { createKbnUrlStateStorage } from '@kbn/kibana-utils-plugin/public';
+import type { DiscoverAppLocatorParams } from '@kbn/discover-plugin/common';
+import type { CoreStart } from '@kbn/core/public';
+import type { DiscoverStart } from '@kbn/discover-plugin/public';
+import { MESSAGE_FIELD, TIMESTAMP_FIELD } from '../../common/constants';
+import {
+  createLogViewStateMachine,
+  DEFAULT_LOG_VIEW,
+  DEFAULT_LOG_VIEW_ID,
+  initializeFromUrl,
+  replaceLogViewInQueryString,
+} from '../observability_logs/log_view_state';
+import { replaceLogFilterInQueryString } from '../observability_logs/log_stream_query_state';
+import { replaceLogPositionInQueryString } from '../observability_logs/log_stream_position_state/src/url_state_storage_service';
+import type { LogsLocatorParams } from './logs_locator';
+import type { InfraClientCoreSetup, QueryTimeRange } from '../types';
+import type { LogViewColumnConfiguration, ResolvedLogView } from '../../common/log_views';
+
+interface LocationToDiscoverParams {
+  core: InfraClientCoreSetup;
+  timeRange?: QueryTimeRange;
+  filter?: string;
+}
+
+export const parseSearchString = ({
+  time,
+  timeRange,
+  filter = '',
+  logViewId = DEFAULT_LOG_VIEW_ID,
+}: LogsLocatorParams) => {
+  return flowRight(
+    replaceLogFilterInQueryString({ language: 'kuery', query: filter }, time, timeRange),
+    replaceLogPositionInQueryString(time),
+    replaceLogViewInQueryString({ type: 'log-view-reference', logViewId })
+  )('');
+};
+
+export const getLocationToDiscover = async ({
+  core,
+  timeRange,
+  filter,
+}: LocationToDiscoverParams) => {
+  const [coreStart, plugins, pluginStart] = await core.getStartServices();
+  const { discover } = plugins;
+  const { logViews } = pluginStart;
+
+  const machine = createLogViewStateMachine({
+    initialContext: { logViewReference: DEFAULT_LOG_VIEW },
+    logViews: logViews.client,
+    initializeFromUrl: createInitializeFromUrl(coreStart),
```

Comment:
It seems dangerous to access the current URL in a locator, since it's intended to be called from within any arbitrary app. Is it possible make the log view reference a parameter here and in the locator so that the query param access happens externally?